guidance/models/transformers/_transformers.py

```diff
@@ -49,81 +53,207 @@ def __init__(
+                    byte_coded = token
+                elif isinstance(token, str):
+                    if hasattr(transformers_tokenizer, "convert_tokens_to_string"):
+                        token_str = transformers_tokenizer.convert_tokens_to_string([token])
+                        encoded_str = transformers_tokenizer.encode(token_str)
+                        if len(encoded_str) != 1:
+                            raise ValueError(f"Round-trip encoding of tokens [{token}] failed! Got {encoded_str}")
+                        roundtrip_id = encoded_str[0]
+                        if roundtrip_id == i:
+                            byte_coded = token_str.encode()
                         else:
-                            byte_coded = token.encode()
+                            byte_coded = bytes([byte_decoder[c] for c in token])
                     else:
-                        raise ValueError(f"Unexpected token type: {type(token)}")
-                byte_tokens[i] = byte_coded
+                        byte_coded = token.encode()
+                else:
+                    raise ValueError(f"Unexpected token type: {type(token)}")
+            byte_tokens[i] = byte_coded
+        return byte_tokens
 
-        else:
-            byte_decoder = transformers_package.AutoTokenizer.from_pretrained(
-                "gpt2", use_fast=False
-            ).byte_decoder  # fall back to gpt2 mapping
+    def _fallback_byte_decoder(self) -> dict[str, int]:
+        byte_decoder = transformers_package.AutoTokenizer.from_pretrained(
+            "gpt2", use_fast=False
+        ).byte_decoder # fall back to gpt2 mapping
+
+        # some special tokens may not have their whitespace encoded...
+        byte_decoder[" "] = 32
+        byte_decoder["\n"] = 10
+        byte_decoder["\r"] = 13
+        byte_decoder["\t"] = 9
+        byte_decoder["▁"] = 32
 
-            # some special tokens may not have their whitespace encoded...
-            byte_decoder[" "] = 32
-            byte_decoder["\n"] = 10
-            byte_decoder["\r"] = 13
-            byte_decoder["\t"] = 9
-            byte_decoder["▁"] = 32
+        return byte_decoder
+
+    def _check_byte_decoder(
+        self,
+        byte_decoder: dict[str, int],
+        transformers_tokenizer: Union[
+            "transformers_package.PreTrainedTokenizer",
+            "transformers_package.PreTrainedTokenizerFast",
+        ],
+    ) -> None:
+
+        def check_byte_decoder_has_all_bytes() -> None:
+            # This is here because some tokenizers are bad and don't have all the bytes (I'm looking at you, microsoft/phi2)
+            all_bytes = set()
+            for x in transformers_tokenizer.get_vocab().keys():
+                for y in x:
+                    all_bytes.add(y)
+            if not set(byte_decoder.keys()) >= all_bytes:
+                raise ByteDecoderError(
+                    f"Byte decoder is missing bytes: {all_bytes - set(byte_decoder.keys())}"
+                )
```

Comment:
This will only get raised if the last-ditch fallback (gpt2 tokenizer) doesn't have all the bytes or fails the complex unicode round-trip (it is raised on line 138 as a `ValueError`). We didn't have any kind of fallback to this in the original code. Really not sure what can be done besides opening an issue here on GH

Reply:
"Please ask the model creator to provide a byte decoder"

Reply:
Almost but not quite. There may have been one on the slow branch that was missing bytes or something, causing the fallback to gpt2.

Reply:
@hudson-ai @Harsha-Nori @riedgar-ms Phi 3 vision is hitting this code path and it's blocking me from progressing. I'll leave a more detailed write up in our teams channel.

Reply:
@nking-1 I believe that the check on the `byte_decoder` here is a sufficient (but not necessary) condition for creating a well-formed set of `byte_tokens`. 

I think you could either proceed by modifying the `byte_decoder` such that it handles spaces/underscores correctly generates the right `byte_tokens` for you (then calling this check just to verify) OR you could figure out how to translate the check on `byte_decoder` to one on the `byte_tokens` themselves... @riedgar-ms any thoughts?